File: KoreanTranslationCamera/KoreanTranslationCamera/Services/OCRService.swift
import Vision
import CoreVideo

// MARK: - OCR Service
//
// Uses Apple's Vision framework (VNRecognizeTextRequest) to detect Korean text
// from a CVPixelBuffer. Only results that contain at least one Hangul character
// are forwarded to the caller; everything else is discarded.
//
// The completion handler is always called on the caller's thread (the camera
// frame queue). Callers are responsible for dispatching to the main queue when
// updating UI.

final class OCRService {

    // MARK: - Public Interface

    func recognizeKoreanText(
        in pixelBuffer: CVPixelBuffer,
        completion: @escaping ([String]) -> Void
    ) {
        let request = VNRecognizeTextRequest { request, _ in
            let candidates = (request.results as? [VNRecognizedTextObservation] ?? [])
                .compactMap { $0.topCandidates(1).first?.string }
                .filter { Self.containsKorean($0) }

            completion(candidates)
        }

        // Korean requires .accurate for reasonable recognition quality.
        request.recognitionLanguages = ["ko-KR", "ko"]
        request.recognitionLevel     = .accurate
        request.usesLanguageCorrection = true

        // Portrait orientation: back camera captures in landscape-right.
        let handler = VNImageRequestHandler(
            cvPixelBuffer: pixelBuffer,
            orientation: .right,
            options: [:]
        )
        try? handler.perform([request])
    }

    // MARK: - Helpers

    private static func containsKorean(_ text: String) -> Bool {
        text.unicodeScalars.contains {
            (0xAC00...0xD7A3).contains($0.value) ||  // Hangul syllables
            (0x1100...0x11FF).contains($0.value) ||  // Hangul Jamo
            (0x3130...0x318F).contains($0.value)     // Hangul Compatibility Jamo
        }
    }
}

File: KoreanTranslationCameraAndroid/app/src/main/java/com/personal/koreantranslation/services/OCRService.kt
package com.personal.koreantranslation.services

import androidx.camera.core.ImageProxy
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.korean.KoreanTextRecognizerOptions

// ---------------------------------------------------------------------------
// OCRService
//
// Uses ML Kit's Korean text recognizer to extract Hangul text blocks from a
// camera frame. Only blocks that contain at least one Korean character are
// forwarded; the rest are discarded.
//
// The callback is invoked on the main thread (ML Kit default).
// The caller must NOT close the ImageProxy before this call completes;
// this service closes it internally after recognition finishes.
// ---------------------------------------------------------------------------

class OCRService {

    private val recognizer = TextRecognition.getClient(
        KoreanTextRecognizerOptions.Builder().build()
    )

    fun recognizeKoreanText(imageProxy: ImageProxy, callback: (List<String>) -> Unit) {
        val mediaImage = imageProxy.image
        if (mediaImage == null) {
            imageProxy.close()
            callback(emptyList())
            return
        }

        val image = InputImage.fromMediaImage(mediaImage, imageProxy.imageInfo.rotationDegrees)

        recognizer.process(image)
            .addOnSuccessListener { visionText ->
                val koreanBlocks = visionText.textBlocks
                    .map { it.text }
                    .filter { containsKorean(it) }
                callback(koreanBlocks)
            }
            .addOnFailureListener {
                callback(emptyList())
            }
            .addOnCompleteListener {
                imageProxy.close()
            }
    }

    fun close() {
        recognizer.close()
    }

    // -----------------------------------------------------------------------

    private fun containsKorean(text: String): Boolean = text.any { ch ->
        ch.code in 0xAC00..0xD7A3 ||  // Hangul syllables
        ch.code in 0x1100..0x11FF ||  // Hangul Jamo
        ch.code in 0x3130..0x318F     // Hangul Compatibility Jamo
    }
}
